### PR TITLE
Text embedder sample for iOS

### PR DESCRIPTION
**Pull request by st-tuanmai**
### Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Fixes # (issue)

### Checklist
Please ensure the following items are complete before submitting a pull request:

- [x] My code follows the code style of the project.
- [ ] I have updated the documentation (if applicable).
- [x] I have added tests to cover my changes.

### Type of Change
Please check the relevant option below:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update (non-breaking change which updates documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Screenshots
If applicable, please add screenshots to help explain your changes.

### Additional Notes
Add any additional information or context about the pull request here.

### CODE REVIEW

**Comment by khanhlvg** [comment]
@priankakariatyml Can you review this app?

File: examples/text_embedder/ios/TextEmbedder/Services/TextEmbedderService.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          let text2Embedder = embeder(text: text2)?.embeddings.first else {
+      return nil
+      
+    }
+     let cosineSimilarity = try? Float(
+      truncating: TextEmbedder.cosineSimilarity(embedding1: text1Embedder, embedding2: text2Embedder))
+    return cosineSimilarity ?? 0
+  }
+
+  func embeder(text: String) -> EmbeddingResult? {
```
Comment (priankakariatyml) [inline examples/text_embedder/ios/TextEmbedder/Services/TextEmbedderService.swift:38]: s/embeder/embed

File: examples/text_embedder/ios/TextEmbedder/Services/TextEmbedderService.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  init(modelPath: String?) {
+    guard let modelPath = modelPath else { return }
+    self.textEmbedder = try? TextEmbedder(modelPath: modelPath)
+  }
+
+  func compare(text1: String?, text2: String?) -> Float? {
+    guard let text1 = text1,
+          let text2 = text2,
+          let text1Embedder = embeder(text: text1)?.embeddings.first,
+          let text2Embedder = embeder(text: text2)?.embeddings.first else {
```
Comment (priankakariatyml) [inline examples/text_embedder/ios/TextEmbedder/Services/TextEmbedderService.swift:29]: text2Embedder/text2EmbedderResult
Similarly for text1

**Comment by priankakariatyml** [comment]
@st-tuanmai  Couple of observations about the UI.
1. The text fields are not readable. Change it to black or something darker.

2. The fact that you have to tap on the background to close the keyboard is not very obvious initially. Compare button is also hidden by the keyboard. I would suggest adding an extra done button on top of the keyboard for dismissal. Can be done using the input accessory view of the text field being edited.

Another way to handle this is putting the content of the view in a scroll view so that users can access `compare` easily. But this will lead to a lot of extra code to properly resize for all phone sizes.

3. UI bug - 
**Steps**:
Tap on text field 1 to edit. Keep the keyboard open (Don't tap on the background to dismiss). -> Tap on text field 2 to edit. 
The UI elements don't move up and show text field 2 completely. (Most of it is hidden).

**Fix**: When the keyboard is already open and you tap on textfield 2, it won't trigger the keyboard frame change notification and hence the UI realignment doesn't happen.  Whatever is done in the keyboard frame change notification, can also be done in `textFieldWillBeginEditing()`. If you have a better idea of where this can be handled, you can choose to do that.

cc: @khanhlvg

File: examples/text_embedder/ios/TextEmbedder/ViewController.swift [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    }
+  }
+
+  // Private function
+  private func setupUI() {
+    input1TextView.text = Constants.defaultText
+    input2TextView.text = Constants.defaultText
+  }
+
+  private func hideKeyboardWhenTappedAround() {
```
Comment (priankakariatyml) [inline examples/text_embedder/ios/TextEmbedder/ViewController.swift:95]: `hideKeyboardOnBackgroundTap`

**Comment by st-tuanmai** [comment]
@priankakariatyml

File: examples/text_embedder/ios/TextEmbedder/ViewController.swift [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    let tap = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
+    tap.cancelsTouchesInView = false
+    view.addGestureRecognizer(tap)
+  }
+
+  private func createInputAccessoryView() -> UIView {
+    let inputAccessoryView = UIView(frame: CGRect(x: 0, y: 0, width: view.bounds.width, height: 44))
+    inputAccessoryView.backgroundColor = .lightGray
+
+    let doneButton = UIButton(frame: CGRect(x: view.bounds.width - 80, y: 5, width: 60, height: 30))
```
Comment (priankakariatyml) [inline examples/text_embedder/ios/TextEmbedder/ViewController.swift:108]: add constants for these values. Like we have in the vision samples. width, spacing from left etc

File: examples/text_embedder/ios/TextEmbedder/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    doneButton.setTitle("Done", for: .normal)
+    doneButton.setTitleColor(.white, for: .normal)
+    doneButton.backgroundColor = UIColor(displayP3Red: 0, green: 127/255.0, blue: 139/255.0, alpha: 1)
```
Comment (priankakariatyml) [inline examples/text_embedder/ios/TextEmbedder/ViewController.swift:111]: Add constant for color

File: examples/text_embedder/ios/TextEmbedder/ViewController.swift [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    clear1Button.isEnabled = !input1TextView.text.isEmpty
+    clear2Button.isEnabled = !input2TextView.text.isEmpty
+    compareButton.isEnabled = !input1TextView.text.isEmpty && !input2TextView.text.isEmpty
+  }
+
+  @objc private func keyboardWillChangeFrame(notification: Notification) {
+    guard let frame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue else { return }
+    if view.bounds.size.height - frame.cgRectValue.origin.y == 0 {
+      contentScrollView.scrollRectToVisible(CGRect(x: 0, y: 0, width: contentScrollView.bounds.width, height: contentScrollView.bounds.height), animated: true)
+      DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
```
Comment (priankakariatyml) [inline examples/text_embedder/ios/TextEmbedder/ViewController.swift:131]: Add constants for the time (0.3, 0.2) etc

**Review by priankakariatyml — APPROVED** [review]

**Pull request merged**